### PR TITLE
improve map integration

### PR DESCRIPTION
**Pull request by timvahlbrock**

### CODE REVIEW

**Comment by timvahlbrock** [comment]
@Paulomart the main issue I'm currently having is, that my useEffect function is called, before the html is first rendered. The width and height of the container will therefore be 0. The map will then only render a single tile, which makes it look like this:

image

Do you have an idea on how to delay the useEffect hook until after the first render?

**Comment by timvahlbrock** [comment]
Issue with map shaking and reset persists, but is eliminated as soon as `useCurrentTime` hook is disabled.

**Comment by timvahlbrock** [comment]
> @Paulomart the main issue I'm currently having is, that my useEffect function is called, before the html is first rendered. The width and height of the container will therefore be 0. The map will then only render a single tile, which makes it look like this:
> 
> image
> Do you have an idea on how to delay the useEffect hook until after the first render?

@Paulomart suggested to use `useLayoutEffect` instead of `setTimeout`, which reduces the likeliness of the problem occurring, but doesn't eliminate it.

File: frontend/src/components/map/LeafletMapContainer.tsx [hunk 1]
```diff
@@ -19,46 +20,46 @@ export interface MapProps {
...
-      return [[currentLocation.details.latitude ?? 51.183334, currentLocation.details.longitude ?? 7.200000]];
+    if (bounds.length > 0) {
+      setView({ bounds });
     }
-    return bounds;
-  }; */
+  }, [locations]);
 
-  const renderMarker = (): ReactElement[] => locations.map((location, index) => <MapMarker key={"marker" + index} origin={origin} destination={destination} location={location} onItemClicked={onItemClicked} />);
+  const renderMarker = (): ReactElement[] =>
```
Comment (Paulomart) [inline frontend/src/components/map/LeafletMapContainer.tsx:44]: no need to use a function for this. just use a const

File: frontend/src/components/map/LeafletMapContainer.tsx [hunk 2]
```diff
@@ -19,46 +20,46 @@ export interface MapProps {
...
+  >
+    <TileLayer
+      attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
+      url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
+    />
+    {renderMarker()}
+    {<CurrentLocationMapMarker />}
+    {showLines
+      ? <Polygon color={"rgb(77, 77, 77)"} opacity={1} dashArray={"20,15"} weight={2} positions={getPolygonPositions()} />
+      : <></>}
```
Comment (Paulomart) [inline frontend/src/components/map/LeafletMapContainer.tsx:61]:
```suggestion
    {showLines &&
      <Polygon color={"rgb(77, 77, 77)"} opacity={1} dashArray={"20,15"} weight={2} positions={getPolygonPositions()} />}
```

File: frontend/src/components/map/LeafletMapContainer.tsx [hunk 3]
```diff
@@ -19,46 +20,46 @@ export interface MapProps {
...
+  return <ReactiveMapContainer
+    style={{ height: "100%", width: "100%" }}
+    view={view}
+  >
+    <TileLayer
+      attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
+      url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
+    />
+    {renderMarker()}
+    {<CurrentLocationMapMarker />}
```
Comment (Paulomart) [inline frontend/src/components/map/LeafletMapContainer.tsx:58]:
```suggestion
    <CurrentLocationMapMarker />
```

File: frontend/src/components/map/LeafletMapContainer.tsx [hunk 4]
```diff
@@ -19,46 +20,46 @@ export interface MapProps {
 const LeafletMapContainer = ({ origin, destination, locationIds, showLines, onItemClicked }: MapProps): JSX.Element => {
   const locations = useMultipleLocationsByIdOrNull(locationIds);
   const usersPosition = useCurrentLocation();
-  const [zoom, setZoom] = useState<number>();
+  const [view, setView] = useState<View>({ bounds: [[50.30527, 5.71687], [52.69499, 9.47241]] });
```
Comment (Paulomart) [inline frontend/src/components/map/LeafletMapContainer.tsx:23]: maybe extract these consts

File: frontend/src/components/map/ReactiveMapContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  }, [props.view]);
+
+  return (<MapContainer
+    style={props.style}
+    ref={mapRef}
+    maxZoom={15}>
+    {props.children}
+  </MapContainer>);
+};
```
Comment (Paulomart) [inline frontend/src/components/map/ReactiveMapContainer.tsx:62]: if you are using ( then please format it nice :)

```suggestion
  return (
    <MapContainer
      style={props.style}
      ref={mapRef}
      maxZoom={15}
    >
      {props.children}
    </MapContainer>
  );
};
```

**Review by Paulomart — CHANGES_REQUESTED** [review]

File: frontend/src/components/map/LeafletMapContainer.tsx [hunk 6]
```diff
@@ -16,49 +17,45 @@ export interface MapProps {
...
+  >
+    <TileLayer
+      attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
+      url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
+    />
+    {markers}
+    <CurrentLocationMapMarker />
+    {showLines
+      ? <Polygon color={"rgb(77, 77, 77)"} opacity={1} dashArray={"20,15"} weight={2} positions={bounds} />
+      : <></>}
```
Comment (Paulomart) [inline frontend/src/components/map/LeafletMapContainer.tsx:57]:
```suggestion
    {showLines && <Polygon color={"rgb(77, 77, 77)"} opacity={1} dashArray={"20,15"} weight={2} positions={bounds} />}
```

Deja Vu?

whoa-deja-vu-matrix-glitch

**Review by Paulomart — CHANGES_REQUESTED** [review]

**Review by Paulomart — APPROVED** [review]

**Pull request merged**